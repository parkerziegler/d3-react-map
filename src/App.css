/** A constants SCSS partial for storing common variables and mixins */
@import url("https://fonts.googleapis.com/css?family=Roboto");
html,
body {
  margin: 0;
  height: 100%;
  width: 100%; }

.root {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  min-height: 100%;
  min-width: 100%; }
  .root .page-container {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    text-align: center;
    min-width: 100%; }
    .root .page-container .page-content {
      display: flex;
      flex-direction: column;
      justify-content: center;
      align-items: center; }

.map-layout {
  display: flex;
  flex-direction: row; }

.chart-layout {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  height: 100%; }

@media only screen and (max-width: 768px) {
  .map-layout {
    display: flex;
    flex-direction: column; } }
