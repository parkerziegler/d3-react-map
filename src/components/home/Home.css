@import url("https://fonts.googleapis.com/css?family=Source+Code+Pro");
.home-text {
  font: 24px "Source Code Pro", monospace; }
