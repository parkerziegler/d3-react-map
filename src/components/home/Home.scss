@import url('https://fonts.googleapis.com/css?family=Source+Code+Pro');

$sourcecode: 'Source Code Pro', monospace;

.home-text {
    font: 24px $sourcecode;
}