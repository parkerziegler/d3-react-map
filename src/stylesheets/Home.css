@import url("https://fonts.googleapis.com/css?family=Source+Code+Pro");
/** A constants SCSS partial for storing common variables and mixins */
@import url("https://fonts.googleapis.com/css?family=Roboto");
/** Use a SCSS loop to create the transition classes to be used
by the react-transition-group add-on */
.home-transition-0-appear {
  opacity: 0.01; }

.home-transition-0-appear.home-transition-0-appear-active {
  opacity: 1;
  transition: opacity 1s ease-in 0s; }

.home-transition-1-appear {
  opacity: 0.01; }

.home-transition-1-appear.home-transition-1-appear-active {
  opacity: 1;
  transition: opacity 1s ease-in 2s; }

.home-transition-2-appear {
  opacity: 0.01; }

.home-transition-2-appear.home-transition-2-appear-active {
  opacity: 1;
  transition: opacity 1s ease-in 4s; }

.home-transition-3-appear {
  opacity: 0.01; }

.home-transition-3-appear.home-transition-3-appear-active {
  opacity: 1;
  transition: opacity 1s ease-in 6s; }

.home-transition-4-appear {
  opacity: 0.01; }

.home-transition-4-appear.home-transition-4-appear-active {
  opacity: 1;
  transition: opacity 1s ease-in 8s; }

.home-transition-5-appear {
  opacity: 0.01; }

.home-transition-5-appear.home-transition-5-appear-active {
  opacity: 1;
  transition: opacity 1s ease-in 10s; }

.home-text {
  font-size: 16px;
  font-family: "Source Code Pro", monospace, sans-serif; }
  @media screen and (min-width: 480px) {
    .home-text {
      font-size: 18px; } }
  @media screen and (min-width: 768px) {
    .home-text {
      font-size: 22px; } }
  @media screen and (min-width: 1024px) {
    .home-text {
      font-size: 28px; } }
  .home-text--large {
    font-size: 16px; }
    @media screen and (min-width: 480px) {
      .home-text--large {
        font-size: 22px; } }
    @media screen and (min-width: 768px) {
      .home-text--large {
        font-size: 28px; } }
    @media screen and (min-width: 1024px) {
      .home-text--large {
        font-size: 36px; } }
