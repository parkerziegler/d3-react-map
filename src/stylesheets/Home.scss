@import url('https://fonts.googleapis.com/css?family=Source+Code+Pro');
@import './_constants.scss';

$sourcecode: 'Source Code Pro', monospace;
$home-font-sizes: (
  null: 16px,
  mobile: 18px,
  tablet: 22px,
  desktop: 28px
);

$large-home-font-sizes: (
  null: 16px,
  mobile: 22px,
  tablet: 28px,
  desktop: 36px
);

/** Use a SCSS loop to create the transition classes to be used
by the react-transition-group add-on */
@for $i from 0 through 5 {
  .home-transition-#{$i}-appear {
    opacity: 0.01;
  }

  .home-transition-#{$i}-appear.home-transition-#{$i}-appear-active {
    opacity: 1;
    transition: opacity 1s ease-in $i * 2s;
  }
}

.home-text {
  @include font-size($home-font-sizes);
  font-family: $sourcecode, sans-serif;
}

.home-text-large {
  @include font-size($large-home-font-sizes);
  font-family: $sourcecode, sans-serif;
}
