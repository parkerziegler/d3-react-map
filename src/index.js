// import necessary modules
import React from 'react';
import ReactDOM from 'react-dom';
import { createStore, compose, applyMiddleware, combineReducers } from 'redux';
import { routerForBrowser, initializeCurrentLocation } from 'redux-little-router';
import { mapReducer } from './reducers/mapReducer';
import { Provider } from 'react-redux';
import App from './App';
import './index.css';
import createSagaMiddleware from 'redux-saga';
import rootSaga from './sagas/rootSaga';

const routes = {
  '/': {
    title: 'Intro',
    index: 0,
    navClass: 'main'
  },
  '/total-shootings': {
    title: 'Total Shootings',
    index: 1,
    navClass: 'main',
    hasChildren: true,
    '/black': {
      title: 'Total Shootings By Race - African American',
      navClass: 'child',
      childIndex: 0,
      hasNextSibling: true
    },
    '/latino': {
      title: 'Total Shootings By Race - Latino',
      navClass: 'child',
      childIndex: 1,
      hasNextSibling: true
    },
    '/white': {
      title: 'Total Shootings By Race - White',
      navClass: 'child',
      childIndex: 2,
      hasNextSibling: false
    }
  },
  '/percapita': {
    title: 'Shootings Per Million By State',
    index: 2,
    navClass: 'main'
  },
  '/shootingsbydate': {
    title: 'Shootings By Date',
    index: 3,
    navClass: 'main',
    isLastRoute: true
  }
};

const {
  reducer,
  middleware,
  enhancer
} = routerForBrowser({
  routes,
  basename: '/d3-react-map'
});

// create our saga middleware - we'll use this handle our side effects
// seamlessly
const sagaMiddleware = createSagaMiddleware();

// create our redux store, applying the saga middleware
const store = createStore(
  combineReducers({ router: reducer, mapReducer }),
  compose(enhancer, applyMiddleware(middleware, sagaMiddleware))
);

// be sure to run the saga middleware
sagaMiddleware.run(rootSaga);

// get the initial location of the router
const initialLocation = store.getState().router;
if (initialLocation) {
  store.dispatch(initializeCurrentLocation(initialLocation));
}

// render our App
ReactDOM.render(
  <Provider store={store}>
    <App />
  </Provider>,
  document.getElementById('root')
);
