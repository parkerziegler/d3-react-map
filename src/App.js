import React, { Component } from 'react';
import { connect } from 'react-redux';
import './App.css';
import Home from './components/home/Home';
import Map from './components/maps/Map/Map';
import MapDescription from './components/maps/MapDescription/MapDescription';
import BarChart from './components/graphs/BarChart/BarChart';
import Chevron from './components/navigation/Chevron/Chevron';
import { callAPI } from './actions/mapActions';
import PropTypes from 'prop-types';
import { Fragment } from 'redux-little-router';
import * as _ from 'lodash';

class App extends Component {

  constructor() {
    super();
    this.findKey = this.findKey.bind(this);
    this.getMainRoutes = this.getMainRoutes.bind(this);
    this.getChildRoutes = this.getChildRoutes.bind(this);
  }

  componentWillMount() {

    // on componentWillMount, invoke our callAPI saga to get our data
    const { dispatch } = this.props;
    dispatch(callAPI());
  }

  findKey(index) {

    // destructure props
    const { router } = this.props;

    // this is just a utility method to find the route index
    // that matches the one we passed in
    return _.findKey(router.routes, (route) => {
      return route.index === index;
    });
  }

  getMainRoutes() {

    // destructure props
    const { router } = this.props;

    let previousMainRoute;
    let nextMainRoute;

    // here we'll need to determine next and previous routes to pass
    // to our Chevron components
    let currentRoute = router.result;

    if (currentRoute.parent) {

      // this is a child route and has no main routes
      // just return '/'
      previousMainRoute = '/';
      nextMainRoute = '/';
    } else {

      // this is a main nav route
      // determine the previous and next routes

      // get the index of the previous main route
      let previousRouteIndex = currentRoute.index - 1;
    
      // add a check to see if we're on the initial route
      previousMainRoute = previousRouteIndex === -1 ? '/' : this.findKey(previousRouteIndex);

      // get the index of the next main route
      let nextRouteIndex = currentRoute.index + 1;

      // add a check to see if this is the last route
      nextMainRoute = currentRoute.isLastRoute ? '/' : this.findKey(nextRouteIndex);
    }

    return {
      previousMainRoute,
      nextMainRoute
    };

  }

  getChildRoutes() {

    const { router } = this.props;

    // here we need to determine previous and next routes
    // for the child navigation (i.e. within a map type)
    let currentRoute = router.result;

    let nextChildRoute;
    let previousChildRoute;

    // if the route has children, it is the parent
    // of the sub-navigation
    if (currentRoute.hasChildren) {

      // the next child route will be the first child
      // and its parent will have the same title as
      // the current route
      nextChildRoute = _.findKey(router.routes, (route) => {

        if (!route.parent) {
          return '/';
        }

        return route.parent.title === currentRoute.title && route.childIndex === 0;
      });

      // there is no previous child route for the parent
      previousChildRoute = '/';

    } else if (currentRoute.hasNextSibling) {

      // if the child route is bordered by other child routes,
      // implement similar logic to the main navigation
      let currentChildRouteIndex = currentRoute.childIndex;

      nextChildRoute = _.findKey(router.routes, (route) => {

        return route.parent.title === currentRoute.parent.title && route.childIndex === currentChildRouteIndex + 1;
      });

      previousChildRoute = currentChildRouteIndex === 0 ? currentRoute.parent.route : _.findKey(router.routes, (route) => {
        return route.parent.title === currentRoute.parent.title && route.childIndex === currentChildRouteIndex - 1;
      });

    } else if (currentRoute.isLastChildRoute) {

      // for the last child route, there is no next child
      let currentChildRouteIndex = currentRoute.childIndex;
      
      nextChildRoute = '/';

      // add a check to see if the previous route is the parent
      // if not, get the previous child route
      previousChildRoute = currentChildRouteIndex === 0 ? currentRoute.parent.route : _.findKey(router.routes, (route) => {
        return route.parent.title === currentRoute.parent.title && route.childIndex === currentChildRouteIndex - 1;
      });

    } else {

      // if this is not a child route, no need to worry about child navigation
      nextChildRoute = '/';
      previousChildRoute = '/';
    }

    return {
      previousChildRoute,
      nextChildRoute
    };

  }

  render() {

    // destructure props
    const { maps, router } = this.props;

    // obtain routes from getMainRoutes and getChildRoutes method
    let { previousMainRoute, nextMainRoute } = this.getMainRoutes();
    let { previousChildRoute, nextChildRoute } = this.getChildRoutes();

    // TODO - implement a Spinner solution
    let component = maps.fetchingData ?
      <div></div> :
      <Fragment forRoute={'/'}>
        <div className='layout'>
          <Chevron className='chevron-link top' path="M1683 1331l-166 165q-19 19-45 19t-45-19l-531-531-531 531q-19 19-45 19t-45-19l-166-165q-19-19-19-45.5t19-45.5l742-741q19-19 45-19t45 19l742 741q19 19 19 45.5t-19 45.5z" href={previousChildRoute} visible={previousChildRoute !== '/'} />
          <div className='page-container'>
            <Chevron className='chevron-link horizontal' path="M1427 301l-531 531 531 531q19 19 19 45t-19 45l-166 166q-19 19-45 19t-45-19l-742-742q-19-19-19-45t19-45l742-742q19-19 45-19t45 19l166 166q19 19 19 45t-19 45z" href={previousMainRoute} visible={previousMainRoute !== '/' || router.route === '/total-shootings'} />
            <div className='page-content'>
              <Fragment forRoute={'/'}>
                <Home />
              </Fragment>
              <Fragment forRoute={'/total-shootings'}>
                <div className='raw'>
                  <Map mapType='proportional' />
                  <MapDescription mapType='proportional' insetHeader='Total Shootings'/>
                </div>
              </Fragment>
              <Fragment forRoute={'/total-shootings/black'}></Fragment>
              <Fragment forRoute={'/total-shootings/latino'}></Fragment>
              <Fragment forRoute={'/total-shootings/asian'}></Fragment>
              <Fragment forRoute={'/total-shootings/nativeamerican'}></Fragment>
              <Fragment forRoute={'/total-shootings/white'}></Fragment>
              <Fragment forRoute={'/percapita'}>
                <div className='percapita'>
                  <Map mapType='choropleth' />
                  <MapDescription mapType='choropleth' insetHeader='Shootings per Million' />
                </div>
              </Fragment>
              <Fragment forRoute={'/shootingsbydate'}>
                  <BarChart />
              </Fragment>
            </div>
            <Chevron className='chevron-link horizontal' path="M1363 877l-742 742q-19 19-45 19t-45-19l-166-166q-19-19-19-45t19-45l531-531-531-531q-19-19-19-45t19-45l166-166q19-19 45-19t45 19l742 742q19 19 19 45t-19 45z" href={nextMainRoute} visible={nextMainRoute !== '/'} />
          </div>
          <Chevron className='chevron-link bottom' path="M1683 808l-742 741q-19 19-45 19t-45-19l-742-741q-19-19-19-45.5t19-45.5l166-165q19-19 45-19t45 19l531 531 531-531q19-19 45-19t45 19l166 165q19 19 19 45.5t-19 45.5z" href={nextChildRoute} visible={nextChildRoute !== '/'} nextSlide={router.routes[nextChildRoute].descSubtitle}/>
        </div>
      </Fragment>;

    return component;
  }
}

const mapStateToProps = (state) => {
  return {
    maps: state.mapReducer,
    router: state.router
  };
};

export default connect(mapStateToProps)(App);

// define propTypes for App
App.propTypes = {
  maps: PropTypes.object.isRequired,
  router: PropTypes.object.isRequired
};

