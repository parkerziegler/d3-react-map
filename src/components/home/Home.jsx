import * as React from 'react';
import { connect } from 'react-redux';
import '../../App.css';
import './Home.css';
import CSSTransitionGroup from 'react-transition-group/CSSTransitionGroup';
import moment from 'moment';

class Home extends React.Component {

    render() {

        const { maps } = this.props;
        
        // first calculate how many shootings occurred per day on average
        let totalCount = Number(maps.shootingsData.length).toLocaleString();
        
        // use moment to get the total number of days in the dataset
        let numDays = moment("2016-12-31").diff(moment("2015-01-01"), "days");

        // get the average number of shootings per data
        let numPerDay = Number(maps.shootingsData.length / numDays).toFixed(1);

        return (
                <div className='page-content' style={{flexDirection: 'column'}}>
                    <CSSTransitionGroup
                    transitionName="header-transition"
                    transitionAppear={true}
                    transitionAppearTimeout={0}
                    transitionEnter={false}
                    transitionLeave={false}>
                        <div className='home-text'>Between January 1, 2015 and December 31, 2016,</div>
                    </CSSTransitionGroup>
                    <CSSTransitionGroup
                    transitionName="header-transition"
                    transitionAppear={true}
                    transitionAppearTimeout={2000}
                    transitionEnter={false}
                    transitionLeave={false}>
                        <div className='home-text' style={{fontSize: 36}}><b>{totalCount}</b></div>
                    </CSSTransitionGroup>
                    <CSSTransitionGroup
                    transitionName="header-transition"
                    transitionAppear={true}
                    transitionAppearTimeout={3500}
                    transitionEnter={false}
                    transitionLeave={false}>
                        <div className='home-text'>people were killed by police in the United States.</div>
                    </CSSTransitionGroup>
                    <CSSTransitionGroup
                    transitionName="header-transition"
                    transitionAppear={true}
                    transitionAppearTimeout={5000}
                    transitionEnter={false}
                    transitionLeave={false}>
                        <div className='home-text'>That amounts to roughly <b>{numPerDay}</b> people per day.</div>
                    </CSSTransitionGroup>
                </div>
        );
    }
}

const mapStateToProps = (state, ownProps) => {
  return {
    maps: state.mapReducer
  };
};

export default connect(mapStateToProps)(Home);
