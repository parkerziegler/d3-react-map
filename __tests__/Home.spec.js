import { store } from '../src/index';
import Home from '../src/components/home/Home';

describe('<Home />', () => {
  it('renders without crashing', () => {
    shallow(<Home store={store} />);
  });

  it('renders all expected child divs', () => {
    const wrapper = shallow(<Home store={store} />);
    expect(
      wrapper
        .dive()
        .find('.page-content')
        .children()
    ).toHaveLength(6);
  });

  it('matches the Jest snapshot', () => {
    const wrapper = mount(<Home store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
    wrapper.unmount();
  });
});
