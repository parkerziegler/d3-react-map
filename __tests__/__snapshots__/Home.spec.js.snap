exports[`<Home /> matches the Jest snapshot 1`] = `"<div class=\"page-content\"><span><div class=\"home-text home-transition-0-appear\">Between January 1, 2015 and December 31, 2016</div></span><span><div class=\"home-text-large home-transition-1-appear\"><b>0</b> people</div></span><span><div class=\"home-text home-transition-2-appear\">were killed by police in the United States.</div></span><span><div class=\"home-text home-transition-3-appear\">That amounts to roughly <b class=\"home-text-large\">0.0</b> people per day.</div></span><span><div class=\"home-text home-transition-4-appear\">Here\'s what we know.</div></span><span><div class=\"home-text home-transition-5-appear\" style=\"font-family: HelveticaNeue, Helvetica, sans-serif; margin-top: 5%; font-weight: 900;\">An Investigation | By Parker Ziegler</div></span></div>"`;
